File: lib/models/match.dart
import 'player.dart';
import 'app_location.dart';
import 'tournament.dart';

class SetScore {
  const SetScore({required this.player1Games, required this.player2Games});

  final int player1Games;
  final int player2Games;

  factory SetScore.fromJson(Map<String, dynamic> json) {
    return SetScore(
      player1Games: json['player1Games'] ?? 0,
      player2Games: json['player2Games'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() => {
    'player1Games': player1Games,
    'player2Games': player2Games,
  };
}

int? scoreWinnerSide(List<SetScore> sets) {
  if (sets.isEmpty) return null;
  var difference = 0;
  for (final set in sets) {
    if (set.player1Games < 0 ||
        set.player2Games < 0 ||
        set.player1Games == set.player2Games) {
      return null;
    }
    difference += set.player1Games > set.player2Games ? 1 : -1;
  }
  return difference == 0
      ? null
      : difference > 0
      ? 1
      : 2;
}

class TennisMatch {
  const TennisMatch({
    required this.id,
    required this.tournament,
    required this.discipline,
    required this.player1,
    required this.player2,
    required this.sets,
    required this.round,
    required this.status,
    required this.friendly,
    this.location,
    this.venue,
    this.scheduledAt,
    this.createdAt,
    this.images = const [],
    this.player1Partner,
    this.player2Partner,
    this.acceptedAt,
    this.resultSubmittedBy,
    this.winner,
    this.statsWinPoints,
  });

  final String id;
  final Tournament? tournament;
  final String discipline;
  final Player player1;
  final Player player2;
  final Player? player1Partner;
  final Player? player2Partner;
  final List<SetScore> sets;
  final Player? winner;
  final int? statsWinPoints;
  final String round;
  final String status;
  final bool friendly;
  final String? location;
  final AppLocation? venue;
  final DateTime? scheduledAt;
  final DateTime? createdAt;
  final List<String> images;
  final DateTime? acceptedAt;
  final Player? resultSubmittedBy;

  String get scoreText =>
      sets.map((set) => '${set.player1Games}-${set.player2Games}').join(', ');

  bool get hasConsistentResult {
    final side = scoreWinnerSide(sets);
    return side != null && winner?.id == (side == 1 ? player1.id : player2.id);
  }

  bool get canRestoreResult =>
      ['disputed', 'rejected', 'cancelled'].contains(status) &&
      hasConsistentResult &&
      resultSubmittedBy != null;

  bool get canReopenResult => [
    'waiting_confirmation',
    'disputed',
    'rejected',
    'cancelled',
    'confirmed',
  ].contains(status);

  bool get requiresLegacyWinPoints =>
      status == 'confirmed' && !friendly && statsWinPoints == null;

  bool get isDoubles => discipline == 'doubles';

  bool includesPlayer(String playerId) =>
      player1.id == playerId ||
      player2.id == playerId ||
      player1Partner?.id == playerId ||
      player2Partner?.id == playerId;

  String get team1Name => player1Partner == null
      ? player1.fullName
      : '${player1.fullName} / ${player1Partner!.fullName}';

  String get team2Name => player2Partner == null
      ? player2.fullName
      : '${player2.fullName} / ${player2Partner!.fullName}';

  factory TennisMatch.fromJson(Map<String, dynamic> json) {
    final venue = json['venue'] is Map<String, dynamic>
        ? AppLocation.fromJson(json['venue'])
        : null;
    return TennisMatch(
      id: (json['_id'] ?? json['id']).toString(),
      tournament: json['tournament'] is Map<String, dynamic>
          ? Tournament.fromJson(json['tournament'])
          : null,
      discipline: json['discipline'] ?? 'singles',
      player1: Player.fromJson(json['player1'] ?? <String, dynamic>{}),
      player2: Player.fromJson(json['player2'] ?? <String, dynamic>{}),
      player1Partner: json['player1Partner'] is Map<String, dynamic>
          ? Player.fromJson(json['player1Partner'])
          : null,
      player2Partner: json['player2Partner'] is Map<String, dynamic>
          ? Player.fromJson(json['player2Partner'])
          : null,
      sets: (json['sets'] as List? ?? [])
          .whereType<Map<String, dynamic>>()
          .map(SetScore.fromJson)
          .toList(),
      winner: json['winner'] is Map<String, dynamic>
          ? Player.fromJson(json['winner'])
          : null,
      statsWinPoints: (json['statsWinPoints'] as num?)?.toInt(),
      round: json['round'] ?? '',
      status: json['status'] ?? 'pending',
      friendly: json['friendly'] == true,
      location: venue?.label ?? json['location'],
      venue: venue,
      scheduledAt: DateTime.tryParse(json['scheduledAt'] ?? ''),
      createdAt: DateTime.tryParse(json['createdAt'] ?? ''),
      images: (json['images'] as List? ?? [])
          .map((item) => item.toString())
          .toList(),
      acceptedAt: DateTime.tryParse(json['acceptedAt'] ?? ''),
      resultSubmittedBy: json['resultSubmittedBy'] is Map<String, dynamic>
          ? Player.fromJson(json['resultSubmittedBy'])
          : null,
    );
  }
}

File: backend/src/models/Match.ts
import { Schema, model, type Types } from "mongoose";

export type MatchStatus =
  | "pending"
  | "accepted"
  | "waiting_confirmation"
  | "confirmed"
  | "rejected"
  | "disputed"
  | "cancelled";

export interface SetScore {
  player1Games: number;
  player2Games: number;
}

export interface MatchAttrs {
  tournament?: Types.ObjectId;
  discipline: "singles" | "doubles";
  player1: Types.ObjectId;
  player2: Types.ObjectId;
  player1Partner?: Types.ObjectId;
  player2Partner?: Types.ObjectId;
  sets: SetScore[];
  winner?: Types.ObjectId;
  round: string;
  location?: string;
  venue?: Types.ObjectId;
  scheduledAt?: Date;
  friendly: boolean;
  status: MatchStatus;
  challengedBy?: Types.ObjectId;
  acceptedAt?: Date;
  resultSubmittedBy?: Types.ObjectId;
  resultSubmittedAt?: Date;
  resultConfirmedBy?: Types.ObjectId;
  confirmedAt?: Date;
  rejectedAt?: Date;
  disputedAt?: Date;
  cancelledAt?: Date;
  adminResolvedBy?: Types.ObjectId;
  adminResolutionNote?: string;
  images: string[];
  statsApplied: boolean;
  statsOperationId?: Types.ObjectId;
  statsWinPoints?: number;
  resultReset?: { id: Types.ObjectId; points: number; actor: Types.ObjectId; note?: string };
  resultHistory: { sets: SetScore[]; winner?: Types.ObjectId; status: MatchStatus; actor: Types.ObjectId; note?: string; changedAt: Date; winPoints?: number }[];
}

const setScoreSchema = new Schema<SetScore>(
  {
    player1Games: { type: Number, required: true, min: 0 },
    player2Games: { type: Number, required: true, min: 0 }
  },
  { _id: false }
);

const matchSchema = new Schema<MatchAttrs>(
  {
    tournament: { type: Schema.Types.ObjectId, ref: "Tournament" },
    discipline: { type: String, enum: ["singles", "doubles"], default: "singles" },
    player1: { type: Schema.Types.ObjectId, ref: "Player", required: true },
    player2: { type: Schema.Types.ObjectId, ref: "Player", required: true },
    player1Partner: { type: Schema.Types.ObjectId, ref: "Player" },
    player2Partner: { type: Schema.Types.ObjectId, ref: "Player" },
    sets: { type: [setScoreSchema], default: [] },
    winner: { type: Schema.Types.ObjectId, ref: "Player" },
    round: { type: String, required: true, trim: true },
    location: { type: String, trim: true },
    venue: { type: Schema.Types.ObjectId, ref: "Location" },
    scheduledAt: { type: Date },
    friendly: { type: Boolean, default: false },
    status: {
      type: String,
      enum: ["pending", "accepted", "waiting_confirmation", "confirmed", "rejected", "disputed", "cancelled"],
      default: "pending"
    },
    challengedBy: { type: Schema.Types.ObjectId, ref: "Player" },
    acceptedAt: { type: Date },
    resultSubmittedBy: { type: Schema.Types.ObjectId, ref: "Player" },
    resultSubmittedAt: { type: Date },
    resultConfirmedBy: { type: Schema.Types.ObjectId, ref: "Player" },
    confirmedAt: { type: Date },
    rejectedAt: { type: Date },
    disputedAt: { type: Date },
    cancelledAt: { type: Date },
    adminResolvedBy: { type: Schema.Types.ObjectId, ref: "Player" },
    adminResolutionNote: { type: String, trim: true },
    images: { type: [String], default: [] },
    statsApplied: { type: Boolean, default: false, select: false },
    statsOperationId: { type: Schema.Types.ObjectId, select: false },
    statsWinPoints: { type: Number, min: 0 },
    resultReset: { type: new Schema({
      id: { type: Schema.Types.ObjectId, required: true },
      points: { type: Number, required: true },
      actor: { type: Schema.Types.ObjectId, required: true },
      note: String
    }, { _id: false }), select: false },
    resultHistory: { type: [new Schema({
      sets: [setScoreSchema], winner: Schema.Types.ObjectId, status: String,
      actor: Schema.Types.ObjectId, note: String, changedAt: Date, winPoints: Number
    }, { _id: false })], default: [], select: false }
  },
  {
    timestamps: true,
    toJSON: {
      transform(_doc, ret) {
        delete (ret as { __v?: number }).__v;
        delete (ret as Partial<MatchAttrs>).statsApplied;
        delete (ret as Partial<MatchAttrs>).statsOperationId;
        delete (ret as Partial<MatchAttrs>).resultReset;
        delete (ret as Partial<MatchAttrs>).resultHistory;
        return ret;
      }
    }
  }
);

matchSchema.index({ tournament: 1, round: 1 });
matchSchema.index({ tournament: 1, player1: 1, player2: 1 });

export const Match = model<MatchAttrs>("Match", matchSchema);
